src/amici.cpp

```diff
@@ -1383,6 +1384,375 @@ int getDiagnosisB(int it, void *ami_mem, UserData *udata, ReturnData *rdata, Tem
+        }
+    }
+    N_VDiv(tdata->xdot, tmp_x, rel_x);
+    rel_res = sqrt(N_VDotProd(rel_x, rel_x));
+    if (res < udata->atol || rel_res < udata->rtol) {
+        // Clean up worksapce
+        N_VDestroy_Serial(delta);
+        N_VDestroy_Serial(rel_x);
+        N_VDestroy_Serial(tmp_x);
+        
+        // Return Success;
+        return (0);
+    }
+    
+    // compute the inital search direction
+    status = getNewtonStep(udata, rdata, tdata, ami_mem, ntry, inewt, delta);
+    
+    // Copy the current state to the old one, make up a new vector for Jdiag
+    N_VScale(1.0, tdata->x, tdata->x_old);
+    N_VScale(1.0, tdata->xdot, tdata->xdot_old);
+    
+    // Newton iterations
+    for(inewt=0; inewt<udata->newton_maxsteps; inewt++) {
+        
+        // Try a full, undamped Newton step
+        N_VLinearSum(1.0, tdata->x_old, gamma, delta, tdata->x);
+        
+        // Ensure positivity of the state
+        x_tmp = N_VGetArrayPointer(tdata->x);
+        for (ix=0; ix<udata->nx; ix++) {
+            if (x_tmp[ix] < 0.0) {
+                x_tmp[ix] = 0.0;
+            }
+        }
+        
+        // Compute new xdot
+        fxdot(tdata->t, tdata->x, tdata->dx, tdata->xdot, udata);
+        
+        // Compute residuals
+        res = N_VDotProd(tdata->xdot, tdata->xdot);
+        res_old = N_VDotProd(tdata->xdot_old, tdata->xdot_old);
+        res = sqrt(res);
+        res_old = sqrt(res_old);
+        
+        // if the step was useful, the dampening can be enlarged
+        if (res<res_old) {
+            // update x and F with x_try and F_try
+            N_VScale(1.0, tdata->x, tdata->x_old);
+            N_VScale(1.0, tdata->xdot, tdata->xdot_old);
+            
+            // Check condition
+            if (res < udata->atol) {
+                // Return number of Newton steps needed
+                rdata->newton_numsteps[ntry-1] = inewt + 1;
+                stat1 = 0;
+                break;
+            }
+            
+            if (stat1 != 0)
+            {
+                // increase dampening
+                gamma = fmax(1.0, 2.0*gamma);
+            
+                // solve the new linear system
+                status = getNewtonStep(udata, rdata, tdata, ami_mem, ntry, inewt, delta);
+            }
+        } else {
+            gamma = gamma/4.0;
+        }
+    }
+
+    // Clean up worksapce
+    N_VDestroy_Serial(delta);
+    N_VDestroy_Serial(rel_x);
+    N_VDestroy_Serial(tmp_x);
+    
+    rdata->newton_numsteps[ntry-1] = udata->newton_maxsteps;
+    
+    return(stat1);
+    
+}
+
+/* ------------------------------------------------------------------------------------- */
+/* ------------------------------------------------------------------------------------- */
+/* ------------------------------------------------------------------------------------- */
+
+int getNewtonStep(UserData *udata, ReturnData *rdata, TempData *tdata, void *ami_mem, int ntry, int nnewt, N_Vector ns_delta) {
```

Comment:
Thanks for pointing out!
I have no clue. I never looked into the one from cvodes (and didn't know it exists) but adapted Glenns code into Amici. For the current commit, I would like to have my routine in Amici, since I know what it does and I don't want bigger changes before the merge. In the very near future, I will have a closer look at the cvodes one and maybe replace mine.